lib/vrl/stdlib/src/decode_mime_q.rs

```diff
@@ -0,0 +1,264 @@
+struct DecodeMimeQFn {
+    value: Box<dyn Expression>,
+}
+
+impl FunctionExpression for DecodeMimeQFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+
+        decode_mime_q(value)
+    }
+
+    fn type_def(&self, _: &state::TypeState) -> TypeDef {
+        TypeDef::bytes().fallible()
+    }
+}
+
+fn decode_mime_q(bytes: Value) -> Resolved {
+    // Parse
+    let input = bytes.try_bytes_utf8_lossy()?;
+    let input: &str = &input;
+    let (remaining, decoded) = alt((
+        fold_many1(
+            parse_delimited_q,
+            || Result::<String>::Ok(String::new()),
+            |result, (head, word)| {
+                let mut result = result?;
+
+                result.push_str(head);
+                result.push_str(&word.decode_word()?);
+
+                Ok(result)
+            },
+        ),
+        alt((
+            map_opt(parse_internal_q, |word| word.decode_word().map(Ok).ok()),
+            success(Ok(String::new())),
+        )),
+    ))(input)
+    .map_err(|e| match e {
+        nom::Err::Error(e) | nom::Err::Failure(e) => {
+            // Create a descriptive error message if possible.
+            nom::error::convert_error(input, e)
+        }
+        nom::Err::Incomplete(_) => e.to_string(),
+    })?;
+
+    let mut decoded = decoded?;
+
+    // Add remaining input to the decoded string.
+    decoded.push_str(remaining);
+
+    Ok(decoded.into())
+}
+
+/// Parses input into (head, (charset, encoding, encoded text))
+fn parse_delimited_q<'a, E: ParseError<&'a str> + ContextError<&'a str>>(
+    input: &'a str,
+) -> IResult<&'a str, (&'a str, EncodedWord<'a>), E> {
+    pair(
+        take_until("=?"),
+        delimited(tag("=?"), parse_internal_q, tag("?=")),
+    )(input)
+}
+
+/// Parses inside of encoded word into (charset, encoding, encoded text)
+fn parse_internal_q<'a, E: ParseError<&'a str> + ContextError<&'a str>>(
+    input: &'a str,
+) -> IResult<&'a str, EncodedWord<'a>, E> {
+    map(
+        separated_pair(
+            opt(take_until1("?")),
+            tag("?"),
+            separated_pair(
+                take_until("?"),
+                tag("?"),
+                alt((take_until("?="), |input| Ok(("", input)))),
+            ),
+        ),
+        |(charset, (encoding, input))| EncodedWord {
+            charset,
+            encoding,
+            input,
+        },
+    )(input)
+}
+
+struct EncodedWord<'a> {
+    charset: Option<&'a str>,
+    encoding: &'a str,
+    input: &'a str,
+}
+
+impl<'a> EncodedWord<'a> {
+    fn decode_word(&self) -> Result<String> {
+        // Modified version from https://github.com/staktrace/mailparse/blob/a83d961fe53fd6504d75ee951a0e91dfea03c830/src/header.rs#L39
+
+        // Decode
+        let decoded = match self.encoding {
+            "B" | "b" => BASE64_MIME
+                .decode(self.input.as_bytes())
+                .map_err(|_| "Unable to decode base64 value")?,
```

Comment:
Does this mean we support both "b" and "q" decoding in this function?

Reply:
Ah, I see the Cue docs call this out:

> Replaces q-encoded or base64-encoded

It's a bit weird, but it's a weird/obscure encoding anyway, so this seems fine to me 👍

Reply:
Yep it's weird. Even that base64 encoding is special version for MIME.